### PR TITLE
common files for TestServer app for download, install and start apps

### PR DESCRIPTION
**Pull request by sridevi-15**
#### Fixes #.

- [x] Ran `flake8`
- [x] Ran `run_repo_tests.sh`

#### Changes proposed in this pull request:

- Files for CBL apps for download, install and start, stop apps

### CODE REVIEW

File: keywords/TestServerBase.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+    def _wait_until_reachable(self, port=None):
+        if not port:
+            port = self.port
+
+        url = "http://{}:{}".format(self.host, port)
+        count = 0
+        while count < MAX_RETRIES:
+            try:
+                self.session.get(url)
```
Comment (hrajput89) [inline keywords/TestServerBase.py]: is this just for check or are we using the response too. If not then we should remove the below commented code.

File: keywords/TestServeriOS.py [hunk 2]
```diff
@@ -0,0 +1,323 @@
...
+
+    def install_device(self):
+        """Installs / launches CBLTestServer on iOS device
+        Warning: Only works with a single device at the moment
+        """
+        if self.debug_mode:
+            self.app_name = "{}-{}-Device-debug.app".format(self.app, self.version_build)
+        else:
+            self.app_name = "{}-{}-Device.app".format(self.app, self.version_build)
+        # package_name = "CBLTestServer-iOS-Device.app"
```
Comment (hrajput89) [inline keywords/TestServeriOS.py]: remove the commented code, if not needed.

File: keywords/TestServeriOS.py [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+
+        if bundle_id not in output:
+            raise LiteServError("Could not install CBLTestServer-iOS")
+
+        self.stop()
+
+    def install(self):
+        """Installs / launches CBLTestServer on iOS simulator
+        """
+        # self.device = "iPhone-7-Plus"
```
Comment (hrajput89) [inline keywords/TestServeriOS.py]: remove the commented code, if not needed.

File: keywords/TestServeriOS.py [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+           The running LiteServ process will be stored in the self.process property.
+        2. The method will poll on the endpoint to make sure LiteServ is available.
+        3. The expected version will be compared with the version reported by http://<host>:<port>
+        4. Return the url of the running LiteServ
+        """
+
+        # self.device = "iPhone-7-Plus"
+        self.logfile_name = logfile_name
+
+        # package_name = "CBLTestServer-iOS.app"
```
Comment (hrajput89) [inline keywords/TestServeriOS.py]: remove the commented code, if not needed.

File: keywords/TestServeriOS.py [hunk 5]
```diff
@@ -0,0 +1,323 @@
...
+        # app_dir = "CBLTestServer-iOS"
+
+        self.app_path = "{}/{}/{}".format(BINARY_DIR, self.app_dir, self.app_name)
+        # self.app_path = "/Users/sridevi.saragadam/Library/Developer/Xcode/DerivedData/CBLTestServer-iOS-ayypbvnmphhpihebbhrjusdqhzmk/Build/Products/Release-EE-iphonesimulator/CBLTestServer-iOS.app"
```
Comment (hrajput89) [inline keywords/TestServeriOS.py]: remove the commented code, if not needed.

File: keywords/TestServeriOS.py [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+
+        output = subprocess.check_output([
+            "ios-deploy", "--justlaunch", "--bundle", self.app_path
+        ])
+        log_info(output)
+
+        self._wait_until_reachable(port=self.port)
+        self._verify_running()
+
+        # return "http://{}:{}".format(self.host, self.port)
```
Comment (hrajput89) [inline keywords/TestServeriOS.py]: remove the commented code, if not needed.

File: keywords/TestServeriOS.py [hunk 7]
```diff
@@ -0,0 +1,323 @@
...
+        # subprocess.Popen didn't launch the app
+        output = subprocess.check_output([
+            "ios-sim", "--devicetypeid", self.device, "launch", self.app_path, "--exit"
+        ])
+
+        log_info(output)
+        self._wait_until_reachable(port=self.port)
+        # time.sleep(10)
+        self._verify_running()
+        # return "http://{}:{}".format(self.host, self.port)
```
Comment (hrajput89) [inline keywords/TestServeriOS.py]: remove the commented code, if not needed.

File: keywords/TestServeriOS.py [hunk 8]
```diff
@@ -0,0 +1,323 @@
...
+
+        return True
+
+    def close_app(self):
+        cur_dir = os.path.dirname(os.path.abspath(__file__))
+        subprocess.check_output(["osascript", "{}/../utilities/sim_close_app.scpt".format(cur_dir)])
+
+    def open_app(self):
+        bundle_id = "com.couchbase.CBLTestServer-iOS"
+        # self.app_path = "/Users/sridevi.saragadam/Library/Developer/Xcode/DerivedData/CBLTestServer-iOS-ayypbvnmphhpihebbhrjusdqhzmk/Build/Products/Debug-iphonesimulator/CBLTestServer-iOS.app"
```
Comment (hrajput89) [inline keywords/TestServeriOS.py]: remove the commented code, if not needed.

File: keywords/couchbaseserver.py [hunk 9]
```diff
@@ -725,33 +725,27 @@ def get_sdk_bucket(self, bucket_name):
...
+         Return the filename portion of the package download URL
+         """
         if version.startswith("3.1.6"):
             return "couchbase-server-enterprise-{}-{}.x86_64.rpm".format(version, cbs_platform)
         elif version.startswith("3.1"):
             return "couchbase-server-enterprise_{}_x86_64_{}-{}-rel.rpm".format(cbs_platform, version, build_number)
         else:
             return "couchbase-server-enterprise-{}-{}-{}.x86_64.rpm".format(version, build_number, cbs_platform)
 
+            return "couchbase-server-enterprise-{}-{}-{}.x86_64.rpm".format(version, build_number, cbs_platform)
```
Comment (hrajput89) [inline keywords/couchbaseserver.py]: duplicate/unreachable return .

File: keywords/TestServerBase.py (hunk 1 above)
Comment (sridevi-15) [reply keywords/TestServerBase.py]: It checks if it throws any connection Error, so do not need response here

File: keywords/TestServerAndroid.py [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+    def download(self, version_build=None):
+        """
+        1. Check to see if .apk is downloaded already. If so, return
+        2. Download the Testserver .apk from latest builds to 'deps/binaries'
+        """
+        if(version_build is not None):
+            self.version_build = version_build
+        version, build = version_and_build(self.version_build)
+
+        # package_name = "CBLTestServer-Android-{}-debug.apk".format(self.version_build)
```
Comment (raghusarangapani) [inline keywords/TestServerAndroid.py]: Can this be removed?

File: keywords/TestServerAndroid.py [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+        output = subprocess.check_output(["adb", "-e", "shell", "pm", "list", "packages"])
+        if "com.couchbase.TestServerApp" not in output:
+            raise LiteServError("Failed to install package: {}".format(output))
+
+        log_info("LiteServ installed to {}".format(self.host))
+
+    def install_device(self):
+        """Install the apk to running Android device or emulator"""
+
+        # apk_name = "CBLTestServer-Android-{}-enterprise-debug.apk".format(self.version_build)
```
Comment (raghusarangapani) [inline keywords/TestServerAndroid.py]: Can this be removed?

File: keywords/TestServerAndroid.py [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+            "--es", "username", "none",
+            "--es", "password", "none",
+            "--ei", "listen_port", str(self.port),
+        ])
+        log_info(output)
+        self._wait_until_reachable(port=self.port)
+        self._verify_launched()
+
+    def _verify_launched(self):
+        """ Poll on expected http://<host>:<port> until it is reachable
```
Comment (raghusarangapani) [inline keywords/TestServerAndroid.py]: This doc string is not right

File: keywords/TestServerAndroid.py [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+
+        log_info("Stopping LiteServ: http://{}:{}".format(self.host, self.port))
+        output = subprocess.check_output(["adb", self.device_option, "shell", "am", "force-stop", "com.couchbase.TestServerApp"])
+        log_info(output)
+
+        # Clear package data
+        output = subprocess.check_output(["adb", self.device_option, "shell", "pm", "clear", "com.couchbase.TestServerApp"])
+        log_info(output)
+
+        # self._verify_not_running()
```
Comment (raghusarangapani) [inline keywords/TestServerAndroid.py]: Does this need to be uncommented?

File: keywords/TestServerFactory.py [hunk 14]
```diff
@@ -0,0 +1,50 @@
...
+    def validate_version_build(version_build):
+        if version_build is None:
+            raise ValueError("Make sure you provide a version / build!")
+
+        if len(version_build.split("-")) != 2:
+            raise ValueError("Make sure your version_build follows the format: 2.0.0-576")
+
+    @staticmethod
+    def validate_platform(platform):
+        valid_platforms = ["android", "ios", "net-mono", "net-msft"]
```
Comment (raghusarangapani) [inline keywords/TestServerFactory.py]: you might want to add net-uwp, xamarin-ios and xamarin-android

File: keywords/TestServeriOS.py [hunk 15]
```diff
@@ -0,0 +1,309 @@
...
+        1. Starts a LiteServ with logging to provided logfile file object.
+           The running LiteServ process will be stored in the self.process property.
+        2. The method will poll on the endpoint to make sure LiteServ is available.
+        3. The expected version will be compared with the version reported by http://<host>:<port>
+        4. Return the url of the running LiteServ
+        """
+
+        self.logfile_name = logfile_name
+        self.app_path = "{}/{}/{}".format(BINARY_DIR, self.app_dir, self.app_name)
+        # self.app_path = "/Users/sridevi.saragadam/Library/Developer/Xcode/DerivedData/CBLTestServer-iOS-ayypbvnmphhpihebbhrjusdqhzmk/Build/Products/Release-EE-iphonesimulator/CBLTestServer-iOS.app"
```
Comment (raghusarangapani) [inline keywords/TestServeriOS.py]: Can this be removed?

File: keywords/TestServeriOS.py (hunk 2 above)
Comment (sridevi-15) [reply keywords/TestServeriOS.py]: done

File: keywords/TestServeriOS.py (hunk 3 above)
Comment (sridevi-15) [reply keywords/TestServeriOS.py]: done

File: keywords/TestServeriOS.py (hunk 4 above)
Comment (sridevi-15) [reply keywords/TestServeriOS.py]: done

File: keywords/TestServeriOS.py (hunk 5 above)
Comment (sridevi-15) [reply keywords/TestServeriOS.py]: done

File: keywords/TestServeriOS.py (hunk 6 above)
Comment (sridevi-15) [reply keywords/TestServeriOS.py]: done

File: keywords/TestServeriOS.py (hunk 7 above)
Comment (sridevi-15) [reply keywords/TestServeriOS.py]: done

File: keywords/TestServeriOS.py (hunk 8 above)
Comment (sridevi-15) [reply keywords/TestServeriOS.py]: done

File: keywords/couchbaseserver.py (hunk 9 above)
Comment (sridevi-15) [reply keywords/couchbaseserver.py]: done

File: keywords/TestServerAndroid.py (hunk 10 above)
Comment (sridevi-15) [reply keywords/TestServerAndroid.py]: done

File: keywords/TestServerAndroid.py (hunk 11 above)
Comment (sridevi-15) [reply keywords/TestServerAndroid.py]: done

File: keywords/TestServerAndroid.py (hunk 12 above)
Comment (sridevi-15) [reply keywords/TestServerAndroid.py]: done

File: keywords/TestServerAndroid.py (hunk 13 above)
Comment (sridevi-15) [reply keywords/TestServerAndroid.py]: I am getting request time out when I try to do http GET, so it is not working as expected, I had to comment.

File: keywords/TestServerFactory.py (hunk 14 above)
Comment (sridevi-15) [reply keywords/TestServerFactory.py]: We should have a seperate PR for that as it needs lot of other changes and other files too.

**Pull request merged**